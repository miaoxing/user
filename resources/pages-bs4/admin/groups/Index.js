import React from "react";
import TableProvider from "components/TableProvider";
import Table from "components/bs4/Table";
import Actions from "components/Actions";
import CDeleteLink from "components/CDeleteLink";
import CEditLink from "components/CEditLink";
import CNewBtn from "components/CNewBtn";
import PageHeader from "components/bs4/PageHeader";
import {Button} from "react-bootstrap4";
import app from "app";
import axios from 'axios';

export default class extends React.Component {
  handleClick = (api) => {
    axios.post(app.url('admin/wechat-groups/sync-from-wechat'), {}, {loading: true})
      .then(({data}) => app.ret(data, api.reload));
  };

  render() {
    return <>
      <TableProvider>
        {api => <>
          <PageHeader>
            <div className="float-right">
              {wei.hasWechatGroup && <Button variant="secondary" onClick={this.handleClick.bind(this, api)}>
                从微信同步分组
              </Button>}
              {' '}
              <CNewBtn/>
            </div>
            {wei.page.controllerTitle}
          </PageHeader>

          <Table
            columns={[
              {
                text: '名称',
                dataField: 'name'
              },
              {
                text: '顺序',
                dataField: 'sort',
                sort: true
              },
              {
                text: '状态',
                dataField: 'wechatId',
                hidden: wei.hasWechatGroup,
                formatter: (cell) => cell > 0 ? '已同步' : '未同步',
              },
              {
                text: '操作',
                formatter: (cell, row) => <Actions>
                  <CEditLink id={row.id}/>
                  <CDeleteLink id={row.id}/>
                </Actions>
              },
            ]}
          />
        </>}
      </TableProvider>
    </>;
  }
}
