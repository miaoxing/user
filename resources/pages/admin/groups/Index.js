import React from "react";
import {Table, TableProvider, CTableDeleteLink} from "@miaoxing/table";
import {CEditLink, CNewBtn} from "@miaoxing/clink";
import {Button} from "react-bootstrap";
import {Page, PageActions} from "@miaoxing/page";
import {LinkActions} from "@miaoxing/actions";
import $ from 'miaoxing';
import api from '@miaoxing/api';

export default class extends React.Component {
  state = {};

  constructor(props) {
    super(props);
    this.ref = React.createRef();
  }

  componentDidMount() {
    api.curPath('metadata', {loading: true}).then(ret => this.setState(ret));
  }

  handleClick = (tableApi) => {
    api.post('wechat-groups/sync-form-wechat', {loading: true}).then(ret => $.ret(ret, tableApi.reload));
  };

  render() {
    return <Page>
      <TableProvider>
        {tableApi => <>
          <PageActions>
            {this.state.hasWechatGroup && <Button variant="secondary" onClick={this.handleClick.bind(this, tableApi)}>
              从微信同步分组
            </Button>}
            <CNewBtn/>
          </PageActions>

          <Table
            columns={[
              {
                title: '名称',
                dataIndex: 'name'
              },
              {
                title: '顺序',
                dataIndex: 'sort',
                sorter: true,
              },
              {
                title: '状态',
                dataIndex: 'wechatId',
                hideInTable: !this.state.hasWechatGroup,
                render: text => text > 0 ? '已同步' : '未同步',
              },
              {
                title: '操作',
                dataIndex: 'id',
                render: (id) => (
                  <LinkActions>
                    <CEditLink id={id}/>
                    <CTableDeleteLink id={id}/>
                  </LinkActions>
                )
              },
            ]}
          />
        </>}
      </TableProvider>
    </Page>;
  }
}
