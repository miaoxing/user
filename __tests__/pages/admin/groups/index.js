import "core-js/stable";
import "regenerator-runtime/runtime";

import React from 'react';
import {MemoryRouter} from 'react-router';
import $ from 'miaoxing';
import Index from '../../../../resources/pages/admin/groups/Index';
import app from '@miaoxing/app';
import {render, waitForElementToBeRemoved, fireEvent} from '@testing-library/react';

function createPromise() {
  let res, rej;

  const promise = new Promise((resolve, reject) => {
    res = (result) => {
      resolve(result);
      return promise;
    };
    rej = (result) => {
      reject(result);
      return promise;
    };
  })

  promise.resolve = res;
  promise.reject = rej;

  return promise;
}

describe('admin/groups', () => {
  beforeAll(() => {
    app.namespace = 'admin';
    app.controller = 'groups';
  });

  afterAll(() => {
    app.namespace = '';
    app.controller = '';
  });

  test('index', async () => {
    const promise = createPromise();
    const promise2 = createPromise();
    const promise3 = createPromise();

    $.get = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        code: 1,
        data: [],
      }))
      .mockImplementationOnce(() => promise2.resolve({
        code: 1
      }))
      .mockImplementationOnce(() => promise3.resolve({
        code: 1,
        data: [
          {
            id: 1,
            name: '测试'
          }
        ]
      }));

    const {container} = render(<MemoryRouter>
      <Index/>
    </MemoryRouter>);

    await Promise.all([promise, promise2, promise3]);

    expect(container.querySelector('.btn-success').innerHTML).toBe('添加');

    expect($.get).toHaveBeenCalledTimes(3);
    expect($.get).toMatchSnapshot();

    await waitForElementToBeRemoved(container.querySelector('.ant-empty'));
  });

  test('delete', async () => {
    // Arrange
    $.get = jest.fn()
      .mockImplementationOnce(() => Promise.resolve({
        code: 1,
        data: [],
      }))
      .mockImplementationOnce(() => Promise.resolve({
        code: 1
      }))
      .mockImplementationOnce(() => Promise.resolve({
        code: 1,
        data: [
          {
            id: 1,
            name: '测试'
          }
        ]
      }));

    // Act 加载页面，可以看到删除按钮
    const {findByText} = render(<MemoryRouter>
      <Index/>
    </MemoryRouter>);

    // Assert
    const link = await findByText('删除');

    // Arrange
    $.confirm = jest.fn().mockImplementationOnce((message, fn) => {
      return fn(true);
    });
    $.post = jest.fn().mockImplementationOnce(() => Promise.resolve({
      code: 1,
    }));
    $.get.mockImplementationOnce(() => Promise.resolve({
      code: 1,
      data: []
    }));

    // Act 点击删除会有弹窗确认
    fireEvent.click(link);

    // Assert
    expect($.confirm).toBeCalledTimes(1);

    // Act Assert 等待刷新，记录不见
    await waitForElementToBeRemoved(link);

    expect($.get).toMatchSnapshot();
    expect($.post).toMatchSnapshot();
  });
});
