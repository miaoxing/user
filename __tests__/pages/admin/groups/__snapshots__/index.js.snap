// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin/groups delete 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "/admin-api/admin-page/breadcrumb",
      Object {
        "params": Object {
          "act": "",
          "ctrl": "admin/groups",
        },
      },
    ],
    Array [
      "/admin-api/groups/metadata",
      Object {
        "loading": true,
      },
    ],
    Array [
      "/admin-api/groups/?page=1&rows=10",
    ],
    Array [
      "/admin-api/groups/?page=1&rows=10",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`admin/groups delete 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "dataType": "json",
        "loading": true,
        "url": "/admin-api/groups/1/destroy",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`admin/groups delete 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "code": 1,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "suc": [Function],
      },
    },
  ],
}
`;

exports[`admin/groups index 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "/admin-api/admin-page/breadcrumb",
      Object {
        "params": Object {
          "act": "",
          "ctrl": "admin/groups",
        },
      },
    ],
    Array [
      "/admin-api/groups/metadata",
      Object {
        "loading": true,
      },
    ],
    Array [
      "/admin-api/groups/?page=1&rows=10",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
